Arrowgene.Ddon.GameServer/Characters/ClanManager.cs

```diff
@@ -0,0 +1,470 @@
+
+            return GetClan(clanId).ClanServerParam.MasterInfo.CharacterListElement.CommunityCharacterBaseInfo.CharacterId;
+        }
+
+        public static CDataClanMemberInfo NewMaster(Character character)
+        {
+            var info = new CDataClanMemberInfo()
+            {
+                Rank = ClanMemberRank.Master,
+                Created = DateTimeOffset.UtcNow,
+                LastLoginTime = DateTimeOffset.UtcNow,
+                Permission = ALL_PERMISSIONS
+            };
+            GameStructure.CDataCharacterListElement(info.CharacterListElement, character);
+            return info;
+        }
+
+        public static CDataClanMemberInfo NewMember(Character character)
+        {
+            var info = new CDataClanMemberInfo()
+            {
+                Rank = ClanMemberRank.Member,
+                Created = DateTimeOffset.UtcNow,
+                LastLoginTime = DateTimeOffset.UtcNow,
+                Permission = 0
+            };
+            GameStructure.CDataCharacterListElement(info.CharacterListElement, character);
+            return info;
+        }
+
+        public List<CDataClanMemberInfo> MemberList(uint clanId)
+        {
+            if (clanId == 0) return new();
+
+            var clan = GetClan(clanId);
+            var memberList = Server.Database.GetClanMemberList(clanId);
+            foreach (var member in memberList)
+            {
+                CDataCommunityCharacterBaseInfo memberInfo = member.CharacterListElement.CommunityCharacterBaseInfo;
+                memberInfo.ClanName = clan.ClanUserParam.ShortName;
+                GameClient lookup = Server.ClientLookup.GetClientByCharacterId(memberInfo.CharacterId);
+                if (lookup != null)
+                {
+                    member.CharacterListElement.OnlineStatus = lookup.Character.OnlineStatus;
+                }
+                else
+                {
+                    member.CharacterListElement.OnlineStatus = OnlineStatus.Offline;
+                }
+            }
+            return memberList;
+        }
+
+        public (uint ClanId, CDataClanMemberInfo MemberInfo) ClanMembership(uint characterId)
+        {
+            var clanId = Server.Database.SelectClanMembershipByCharacterId(characterId);
+            var membership = Server.Database.GetClanMember(characterId);
+
+            return (clanId, membership);
+        }
+
+        public void SendToClan<T>(uint clanId, T packet)
+            where T : class, IPacketStructure, new()
+        {
+            foreach (var client in Server.ClientLookup.GetAll())
+            {
+                if (client.Character != null && client.Character.ClanId == clanId)
+                {
+                    client.Send(packet);
+                }
+            }
+        }
+
+        private static int BitsToInt(IEnumerable<int> bitindices)
+        {
+            return bitindices.Aggregate(0, (sum, val) => sum + 2 ^ (val));
```

Comment:
What is this doing exactly?

Reply:
Sure, but I am unsure of the math. Why is it `sum + (2^val)`. Normally you would just take a bitmask and do a bitwise and on the value to see if it is set.